### ISSUE TITLE
How to deal padding in LSTM and Loss calculation?

### ISSUE THREAD

**Issue by WenChouWu** [2019-dl-training-program/Lab4#10]
Hello TA,

In your last supplement, you mentioned that we could use `nn.CrossEntropyLoss(ignore_index=padding_index)` to ignore the loss calculation of padding vector. Does this mean that we don't need to remove the padding vector, like what we do with pack_padded_sequence with batches, when we input the embedded vector to GRU/LSTM? Am I correct?

**Comment by robert80203** [comment]
ignore_index only solve the padding problem in the decoder part.
In the encoder part, if you seq_len is not 1, you will still face padding problem
If you adopt ignore_index=padding, that means in the decoder part, you don't have to break the loop since the padding output wiil not be included in the loss.
My suggestion here is that you should use the simple training method at first (loop with decoder without using ignore_index). There is no padding problem when you use looping method as you will break the loop when you face EOS token

**Comment by WenChouWu** [comment]
Seems there is something wrong that I can not reduce the reconstruction loss even I set the KLD weighting = 0. Just want to confirm if I do the LSTM correctly.

My code is similar to below one and the major idea is to input variable length of sequences into GRU/LSTM. Do this match your suggestion?

```
word1 = torch.LongTensor([1,2,3,4,5,6,7,8,9,10]).view(-1,1)
word2 = torch.LongTensor([11,12,13,14,15,0,0,0,0,0]).view(-1,1)
word3 = torch.LongTensor([16,17,0,0,0,0,0,0,0,0]).view(-1,1)
word = torch.cat((word1, word2, word3), dim=1)

embed = nn.Embedding(20,256)
word_embed = embed(word)

nnLSTM = nn.LSTM(256,512)

output, hidden = nnLSTM(word_embed[:,0,:][:10,:].unsqueeze(1))  # word1, length=10
output, hidden = nnLSTM(word_embed[:,1,:][:5,:].unsqueeze(1), hidden)  # word2, length=5
output, hidden = nnLSTM(word_embed[:,2,:][:2,:].unsqueeze(1), hidden)  # word3, length=2
```

**Comment by robert80203** [comment]
It seems your variable length sequence in nothing wrong. Perhaps you should check the values of your outputs and weights to see whether they are reasonable and make sure you average the loss.
